### PR TITLE
[core] Introduce Context Switching Mechanism

### PR DESCRIPTION
**Pull request by joaovicentesouto**
## Description

In this PR, I introduce the interface to create and exchange execution contexts. Besides the default interface, the incoming changes already bring the implementation for MPPA-256 architecture.

This feature is verified by two unit tests:
- Creation of a context
- Context switching and restoration using three levels of contexts.

## Related issue

Mention #620

### CODE REVIEW

File: src/hal/arch/core/k1b/ctx.c [hunk 1]
```diff
@@ -26,8 +26,16 @@
 #define __NEED_CORE_CONTEXT
 
 #include <arch/core/k1b/ctx.h>
+#include <arch/core/k1b/mmu.h>
+#include <arch/core/k1b/mOS.h>
 #include <nanvix/const.h>
 #include <nanvix/hlib.h>
+#include <HAL/hal/core/legacy.h>
```
Comment (ppenna) [inline src/hal/arch/core/k1b/ctx.c:33]: Move this to `<arch/core/k1b/mOS.h>`

File: src/hal/arch/core/k1b/ctx.c [hunk 2]
```diff
@@ -61,3 +69,79 @@ PUBLIC void k1b_context_dump(const struct context *ctx)
...
+	ps.ie   = 1;
+	ps.hle  = 1;
+	ps.ice  = 1;
+	ps.use  = 1;
+	ps.dce  = 1;
+	ps.isw  = 1; //! Use kernel stack on interrupts.
+	ps.esw  = 1; //! Use kernel stack on traps.
+
+	/* Sets magic number. */
+	ustack->magic = 0xdeadbeef;
```
Comment (ppenna) [inline src/hal/arch/core/k1b/ctx.c:108]: Constant for `0xdeadbeef` ?

File: src/hal/arch/core/k1b/ctx.c [hunk 3]
```diff
@@ -61,3 +69,79 @@ PUBLIC void k1b_context_dump(const struct context *ctx)
...
+
+	/* Populate context. */
+	ctx->ps  = ps.word;        /**< Processing Status.                           */
+	ctx->sps = ps.word;        /**< Shadown Processing Status.                   */
+	ctx->spc = (word_t) start; /**< Start routine called.                        */
+	ctx->ssp = (word_t) sp;    /**< User stack prepared to call start routine.   */
+	ctx->ksp = (word_t) ksp;   /**< Kernel stack pointer.                        */
+	ctx->r12 = (word_t) ctx;   /**< Kernel stack where context will be consumed. */
+	ctx->r13 = (word_t) bp;    /**< Base stack pointer.                          */
+	ctx->cs  = 0x8000;         /**< Reset value.                                 */
```
Comment (ppenna) [inline src/hal/arch/core/k1b/ctx.c:138]: What `0x8000` means?

File: src/hal/arch/core/k1b/ctx.c [hunk 4]
```diff
@@ -61,3 +69,79 @@ PUBLIC void k1b_context_dump(const struct context *ctx)
...
+
+	KASSERT(((word_t) ctx) % 8 == 0);
+	KASSERT(((word_t) ctx->ssp) % 8 == 0);
+	KASSERT(((word_t) ctx->ksp) % 8 == 0);
+	KASSERT(((word_t) ctx->r12) % 8 == 0);
```
Comment (ppenna) [inline src/hal/arch/core/k1b/ctx.c:143]: `&7 goes faster`

File: src/hal/core/context.c [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#if 0
+	if (start != text_region)
+		return (NULL);
+
+	if (ustack != user_space || (ustack + USTACK_SIZE) != user_space)
+		return (NULL);
+
+	if (kstack != kernel_space || (kstack + KSTACK_SIZE) != kernel_space))
+		return (NULL);
+#endif
```
Comment (ppenna) [inline src/hal/core/context.c:66]: Dead code?

File: src/hal/core/context.c [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+	/* Is the next pointer invalid? */
+	if (next == NULL || (*next) == NULL)
+		return (-EINVAL);
+
+#if 0
+	/* Is pointer stored into user space? */
+	if (*next != kernel_space)
+		kpanic("[core][restore] Invalid context: user space context!");
+#endif
```
Comment (ppenna) [inline src/hal/core/context.c:108]: Dead code?

**Comment by ppenna** [comment]
Awesome work btw

File: src/hal/arch/core/k1b/ctx.c (hunk 1 above)
Comment (joaovicentesouto) [reply src/hal/arch/core/k1b/ctx.c]: Moved.

File: src/hal/arch/core/k1b/ctx.c (hunk 2 above)
Comment (joaovicentesouto) [reply src/hal/arch/core/k1b/ctx.c]: Defined.

File: src/hal/arch/core/k1b/ctx.c (hunk 3 above)
Comment (joaovicentesouto) [reply src/hal/arch/core/k1b/ctx.c]: This value is the reset value required by the hardware. It is specified in the Kalray's manual.

File: src/hal/arch/core/k1b/ctx.c (hunk 4 above)
Comment (joaovicentesouto) [reply src/hal/arch/core/k1b/ctx.c]: Done.

File: src/hal/core/context.c (hunk 5 above)
Comment (joaovicentesouto) [reply src/hal/core/context.c]: Removed.

File: src/hal/core/context.c (hunk 6 above)
Comment (joaovicentesouto) [reply src/hal/core/context.c]: Removed.

**Review by ppenna — APPROVED** [review]

**Pull request merged**